crates/oxc_codegen/src/sourcemap_builder.rs

```diff
@@ -131,12 +136,29 @@ impl SourcemapBuilder {
 
     #[allow(clippy::cast_possible_truncation)]
     fn search_original_line_and_column(&mut self, position: u32) -> (u32, u32) {
-        let result = self
-            .line_offset_tables
-            .lines
-            .partition_point(|table| table.byte_offset_to_start_of_line <= position);
-        let original_line = if result > 0 { result - 1 } else { 0 };
-        let line = &self.line_offset_tables.lines[original_line];
+        let lines = &self.line_offset_tables.lines;
+        let mut idx = self.last_line_lookup;
+
+        // Ensure we have a valid starting point
+        // If position < lines[idx].byte_offset_to_start_of_line, fallback to a full binary search.
+        if position < lines[idx].byte_offset_to_start_of_line {
+            #[cold]
+            fn fallback_binary_search(lines: &[Line], position: u32) -> usize {
+                let mut i =
+                    lines.partition_point(|line| line.byte_offset_to_start_of_line <= position);
+                i = i.saturating_sub(1);
+                i
+            }
+            idx = fallback_binary_search(lines, position);
+        } else {
+            while idx + 1 < lines.len() && lines[idx + 1].byte_offset_to_start_of_line <= position {
+                idx += 1;
+            }
+        }
```

Comment:
I'm not sure if `#[cold]` is a good idea here or not.

In our benchmark, it parses a file and then prints it and generates sourcemap, without any alteration to the AST in between. So the order of code in output is always the same as order of input. So the fallback branch is 100% *never* taken in the benchmark.

But that's probably unrealistic. Usually you'd modify the AST in some way between parse and print (otherwise, what's the point?). And it's not uncommon for transformer to move code to be in different order from source, which would then hit the fallback branch.

Probably we should change the benchmark to run the AST through transformer first, to make the benchmark more realistic. Would you like to do that, and then stack this PR on top, to make sure it still has the positive effect we're seeing now?

Also, could you possibly try the following, just to see if it makes a difference either way:

1. Remove the `fallback_binary_search` function. Just keep the fallback code inline.
2. Reverse the `if` arms. i.e. make the condition `if position >= lines[idx].byte_offset_to_start_of_line`.

Reason for reversing the `if` arms is that *usually* compiler will order the generated assembly on assumption that the `if` branch is taken more often than the `else` branch. So it's a "softer" way of hinting what branch you consider more likely to be taken than using `#[cold]`.

To be honest, I don't entirely know what `#[cold]` does. I get the general idea of it, but don't know the subtleties of exactly how it interacts with LLVM and branch predictor. So I'm suggesting this as an experiment, rather than "this will definitely work". But it'd be interesting to try if you're willing and have time.

If you push those changes as a separate commit, then CodSpeed will show you a relative change before/after the change.

Reply:
> But that's probably unrealistic. Usually you'd modify the AST in some way between parse and print (otherwise, what's the point?). And it's not uncommon for transformer to move code to be in different order from source, which would then hit the fallback branch.
> Probably we should change the benchmark to run the AST through transformer first, to make the benchmark more realistic. Would you like to do that, and then stack this PR on top, to make sure it still has the positive effect we're seeing now?

yeah this is a very good point. i've put up a PR #7934 to achieve this. Lets see what the benchmarks say 🙂

Reply:
so this cold branch is incredibly rare to be hit. (i tried this with the transformer and enable all)

I've been benching this against `checker.ts` ( i copied the main function 2x to make the file a bit bigger)
But as you can see, the edge case is only hit 16 times.
Furthermore, the diff indicates that we probably don't need binary search here.
`-10, -23, -4, -5, -10, -4, -5, -5, -4, -5, -3, -3, -8, -5, -4, -6` (this represents how many lines we went back to find what we were looking for.
```
[crates/oxc_codegen/src/sourcemap_builder.rs:105:9] self.total_search_count = 353513
[crates/oxc_codegen/src/sourcemap_builder.rs:105:9] self.total_search_fallback_count = 16
```

1. Removing `#[cold]`

| case             | time |
| ---------------- | ---- |
| `main`           | `33.63`    |
| current PR state | `28.54`    |
| remove `#[cold]` | `28.32`    |

This is average over 100 runs. i did this twice, so looks like no `#[cold]` very marginally faster.

2. Reversing `if` conditions:
| case             | time |
| ---------------- | ---- |
| `main`           | `33.78`    |
| current PR state | `29.09`    |
| reversing if conditions | `28.64`    |
| reverse `if` AND remove `#[cold]` | |

^^ Note for the above, the diff between in the current state/reversing conditions was increcibly flaky about which one was consistently fast

i've put each of these changes in a separate commit. lets see what codespeed says

Reply:
looks like there's virtually no difference on codespeed

Reply:
Excellent research!

Yes, it looks avoiding binary search for the fallback case is a good idea too.

But... I do have one concern. Oxc is designed for general use. Users can perform *any* transform on AST that they like.

As an example, they might choose to hoist all function declarations to the top of the file. That would mean that a function which was previously at the very bottom of a file is now at the very top.

So, if the file is large, using linear search only would be pathologically slow.

Here's a suggestion:

* In both directions, do linear search for `N` lines.
* If not found, fallback to binary search.

I imagine that'd keep the perf gain of this PR in the common case, while avoiding danger of extremely bad perf where code has moved a long distance.

Taking a slice of `lines` before the linear search loop, and iterating over that slice *might* even make it faster, by avoiding a bounds check on each turn of the loop. Maybe.

I'm not sure what a good value for `N` is. From your numbers above, I'd guess 16 is good for reverse search (it'd only fallback to binary search once), but not sure if 16 is good for forwards too.

By the way, we haven't yet enabled class transforms (still working on them) and they move code around a *lot* ([Babel REPL](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=MYGwhgzhAEDC0G8BQ1oHo3QCoAsCWMEALgK4BmZ0YII0AtgPYBuApjEQ1WUSwE7REcLaKEgQU0YmCJ5g0AA68G86AF5oZBgwAUASgDcEgMSK8TaSz2IAvhKky5yVKgBGYXnsOpbqCfJIuILJWTs4Y0AAyDEQwDJTADAAmwkK8LBK2EuG4BJKkFPTMbAKcLiyaaWgcKnECQiLgUBIA2gCOJAAeegC6atCBYDgDOJ5ItkA&debug=false&forceAllTransforms=false&modules=false&shippedProposals=false&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=&prettier=true&targets=&version=7.26.4&externalPlugins=%40babel%2Fplugin-transform-class-properties%407.25.9%2C%40babel%2Fplugin-external-helpers%407.25.9%2C%40babel%2Fplugin-transform-private-methods%407.25.9%2C%40babel%2Fplugin-transform-class-static-block%407.26.0&assumptions=%7B%7D)).

Reply:
CodSpeed has a blind spot - it does not take branch prediction hits/misses into account in its measures. So your local criterion benchmarks are a better measure when assessing whether `#[cold]` and `if` arms reversal are good or not.

This is a major shortcoming in our benchmarking infrastructure which I hope we can remedy at some point. https://github.com/oxc-project/backlog/issues/5

Reply:
this makes a lot of sense. i've set `N` to 20. we search back/fwd 20 lines, else fall back to bin search

Reply:
I suggested 16 as it might encourage compiler to use SIMD (compiler loves powers of 2). But maybe it wouldn't do that anyway.

Reply:
> Merging https://github.com/oxc-project/oxc/pull/7926 will not alter performance

rip, back to square one